src/source.cpp

```diff
@@ -79,6 +90,121 @@ unique_ptr<Source> Source::create(pugi::xml_node node)
+    auto ids = get_node_array<double>(node, "time_bounds");
+    if (ids.size() == 2)
+      time_bounds_ = std::make_pair(ids[0], ids[1]);
+  }
+  if (check_for_node(node, "energy_bounds")) {
+    auto ids = get_node_array<double>(node, "energy_bounds");
+    if (ids.size() == 2)
+      energy_bounds_ = std::make_pair(ids[0], ids[1]);
+  }
+
+  // Check for how to handle rejected particles
+  if (check_for_node(node, "rejection_strategy")) {
+    std::string rejection_strategy = get_node_value(node, "rejection_strategy");
+    if (rejection_strategy == "kill") {
+      rejection_strategy_ = RejectionStrategy::KILL;
+    } else if (rejection_strategy == "resample") {
+      rejection_strategy_ = RejectionStrategy::RESAMPLE;
+    } else {
+      fatal_error(std::string(
+        "Unrecognized strategy source rejection: " + rejection_strategy));
+    }
+  } else {
+    // Default to resample rejected particles
+    rejection_strategy_ = RejectionStrategy::RESAMPLE;
+  }
+}
+
+SourceSite Source::sample(uint64_t* seed) const
+{
+  bool found = false;
+  SourceSite site;
+
+  while (!found) {
+    // Sample a particle randomly from list
+    site = this->sample_no_rejection(seed);
+
+    found = satisfies_constraints(site);
+    if (!found && rejection_strategy_ == RejectionStrategy::KILL) {
+      // Accept particle regardless but set wgt=0 to trigger garbage collection
+      // I.e. kill this particle immediately and pick the next
+      found = true;
+      site.wgt = 0.0;
+    }
+  }
+  return site;
+}
+
+bool Source::satisfies_constraints(SourceSite& s) const
+{
+  return satisfies_spatial_constraints(s.r) &&
+         satisfies_energy_constraints(s.E) &&
+         satisfies_time_constraints(s.time);
+}
+
+bool Source::satisfies_energy_constraints(double E) const
+{
+  return E > energy_bounds_.first && E < energy_bounds_.second;
+}
+
+bool Source::satisfies_time_constraints(const double time) const
+{
+  return time > time_bounds_.first && time < time_bounds_.second;
+}
+
+bool Source::satisfies_spatial_constraints(Position r) const
+{
+  bool found = false;
+  GeometryState geom_state;
+  geom_state.r() = r;
+
+  // Reject particle if it's not in the geometry at all
+  if (!(found = exhaustive_find_cell(geom_state)))
+    return false;
+
+  if (!domain_ids_.empty()) {
+    if (domain_type_ == DomainType::MATERIAL) {
+      auto mat_index = geom_state.material();
+      if (mat_index != MATERIAL_VOID) {
+        found = contains(domain_ids_, model::materials[mat_index]->id());
+      }
+    } else {
+      for (int i = 0; i < geom_state.n_coord(); i++) {
+        auto id = (domain_type_ == DomainType::CELL)
+                    ? model::cells[geom_state.coord(i).cell]->id_
+                    : model::universes[geom_state.coord(i).universe]->id_;
+        if ((found = contains(domain_ids_, id)))
+          break;
+      }
+    }
+  }
+  // If at this point found is false, part. is outside all cells/materials
```

Comment:
Let's not abbreviate `particle`. I was thinking this was a typo period and referring to a CAD part!